appliance/mariadb/process.go

```diff
@@ -0,0 +1,870 @@
+					logger.Error("error stopping process", "err", err)
+				}
+				return err
+			default:
+				logger.Debug("ignoring error connecting to mysql", "err", err)
+				time.Sleep(checkInterval)
+				continue
+			}
+		}
+
+		return nil
+	}
+}
+
+func (p *Process) stop() error {
+	logger := p.Logger.New("fn", "stop")
+	logger.Info("stopping mysql")
+
+	p.cancelSyncWait()
+
+	// Attempt to kill.
+	logger.Debug("stopping daemon")
+	if err := p.cmd.Stop(); err != nil {
+		logger.Error("error stopping command", "err", err)
+	}
+
+	// Wait for cmd to stop or timeout.
+	select {
+	case <-time.After(p.OpTimeout):
+		return errors.New("unable to kill process")
+	case <-p.cmd.Stopped():
+		p.runningValue.Store(false)
+		return nil
+	}
+}
+
+func (p *Process) Info() (*ProcessInfo, error) {
+	return &ProcessInfo{
+		Config:           p.config(),
+		Running:          p.running(),
+		SyncedDownstream: p.syncedDownstream(),
+	}, nil
+}
+
+func (p *Process) waitForSync(downstream *discoverd.Instance, enableWrites bool) {
+	stopCh := make(chan struct{})
+	doneCh := make(chan struct{})
+
+	var once sync.Once
+	p.cancelSyncWait = func() {
+		once.Do(func() { close(stopCh); <-doneCh })
+	}
+
+	go func() {
+		defer close(doneCh)
+
+		startTime := time.Now().UTC()
+		logger := p.Logger.New(
+			"fn", "waitForSync",
+			"sync_name", downstream.Meta["MYSQL_ID"],
+			"start_time", log15.Lazy{func() time.Time { return startTime }},
+		)
+
+		logger.Info("waiting for downstream replication to catch up")
+		defer logger.Info("finished waiting for downstream replication")
+
+		for {
+			// Check if "wait sync" has been canceled.
+			select {
+			case <-stopCh:
+				logger.Debug("canceled, stopping")
+				return
+			default:
+			}
+
+			// Read local master status.
+			masterXLog, err := p.XLogPosition()
+			if err != nil {
+				logger.Error("error reading master xlog", "err", err)
+				startTime = time.Now().UTC()
+				select {
+				case <-stopCh:
+					logger.Debug("canceled, stopping")
+					return
+				case <-time.After(checkInterval):
+				}
+				continue
+			}
+			logger.Info("master xlog", "gtid", masterXLog)
+
+			// Read downstream slave status.
+			slaveXLog, err := p.getXLogPosition(&DSN{
+				Host:     downstream.Addr,
+				User:     "flynn",
+				Password: p.Password,
+				Timeout:  p.OpTimeout,
+			})
+			if err != nil {
+				logger.Error("error reading slave xlog", "err", err)
+				startTime = time.Now().UTC()
+				select {
+				case <-stopCh:
+					logger.Debug("canceled, stopping")
+					return
+				case <-time.After(checkInterval):
+				}
+				continue
+			}
+
+			logger.Info("mysql slave xlog", "gtid", slaveXLog)
+
+			elapsedTime := time.Since(startTime)
+			logger := logger.New(
+				"master_log_pos", masterXLog,
+				"slave_log_pos", slaveXLog,
+				"elapsed", elapsedTime,
+			)
+
+			// Update the slave status and reset start time if replication has progressed.
+			// FIXME(jpg): What is the intention here? To track slave xlog over each
```

Comment:
Yeah, I think this is copied from the Postgres code which is in turn copied from the upstream Manatee code.